---
title: "Fig2_extdataFig2"
output: html_document
---


**Load the 'treated_samples.rds' object:**

``` 
  treated_samples<-readRDS(file = file.path(base_dir, "treated_samples.rds"))
```

**Fig.2a:**Readthrough distribution across drugs.
```
  ggplot(treated_samples[!is.na(RT)],aes(x = RT, y = treatment)) +
    geom_density_ridges(aes(color=treatment, scale = 2, fill=treatment)) +
    # basal_RT==T layer: geom_density_ridges(data=treated_samples[basal_RT=="high" & replicate==1], alpha=0.1, fill="orange", aes(color=treatment, scale = 2, fill=treatment)) + 
    # geom_jitter(data = treated_samples[basal_RT=="high" & replicate==1], aes(y=treatment, x=RT), width = 0.0, height = 0.02) +
    scale_fill_viridis_d(alpha=0.75) +
    scale_color_viridis_d() +
    scale_y_discrete(expand = c(0, 0), limits=c("Untreated", "Gentamicin", "FUr", "CC90009", "G418", "Clitocine", "DAP", "SJ6986", "SRI"))+
    scale_x_continuous(limits = c(-0.25, 7.5), breaks=seq(1,7.5,1))+
    ylab("Density")+
    xlab("Readthrough (%)")+ geom_vline(xintercept=1,linetype="dotted")+
    theme_classic()+
    xlim(0,5) +
    theme(axis.text.y = element_text(vjust = 0.5,
                                     hjust = 0.5,
                                     angle = 0),
          axis.text.x = element_text(vjust = 0.5,
                                     hjust = 0.5,
                                     angle = 0),
          legend.title = element_blank(),
          legend.position = "none",
          text = element_text(size = 20),
          legend.text = element_text(size=12)) +
    labs(title="") + p
```


     
 
**Fig.2b**: Interdrugs correlations. Load the correlations_tbl, cluster and plot the heatmap:
```
  correlations_tbl<-readRDS(file = "~/R/RT/Disease_PTCs/Disease_objects/correlations_tbl.rds")
```
Cluster the correlations_tbl to make the heatmap more readable:

  'r' is a rearrangement of correlations_tbl, to do the clustering. 
```  
  r<-tidyr::pivot_wider(correlations_tbl, names_from = "treatment_2", values_from = "correlation")
  r <- as.matrix(r[, -1]) # remove the treatment_1 column to do the clustering (we don't want it, it would mess things up)
  clust_rows_heatmap2 <- hclust(dist(t(r)))
```
  Do the clustering for the columns 
  
```
  c<-tidyr::pivot_wider(correlations_tbl, names_from = "treatment_2", values_from = "correlation")
  c <- as.matrix(c[, -1]) # remove the variant column to do the clustering (we don't want it, it would mess things up)
  clust_cols_heatmap2 <- hclust(dist(t(c)))
```

Alternatively, directly load the 'correlations_tbl.rds' object from base_dir:

```
correlations_tbl<-readRDS(file = file.path(base_dir, "Fig2/correlations_tbl.rds"))
```

  Plot:
```
 ggplot(correlations_tbl, aes(x=treatment_1 ,y=treatment_2)) + geom_tile(aes(fill=correlation))+scale_fill_gradient2(low="red",mid="white",high="blue",midpoint=0, limits = c(0, 1), oob = scales::squish) + 
          labs(fill="Pearson\ncorrelation") + scale_y_discrete(limits = colnames(r)[clust_rows_heatmap2$order]) + scale_x_discrete(limits=colnames(c)[clust_cols_heatmap2$order]) +
          p + theme(text = element_text(size = 15), axis.text.x = element_text(size=15, angle=90), axis.text.y = element_text(size=15), axis.title.x = element_blank(), axis.title.y = element_blank())
    
```
  
**Fig.2b:** scatter plots.
Generate the 'ax' datatables each contains data for one drug.

```
  a1<-treated_samples[treatment=="Untreated" & replicate==1, c("identifier","RT", "stop_type", "down_12nt", "down_1nt")]; setnames(a1, "RT", "Untreated")
  
  a2<-treated_samples[treatment=="SRI" & replicate==1, c("identifier","RT")]; setnames(a2, "RT", "SRI")
  a3<-treated_samples[treatment=="DAP" & replicate==1, c("identifier","RT")]; setnames(a3, "RT", "DAP")
  a4<-treated_samples[treatment=="Clitocine" & replicate==1, c("identifier","RT")]; setnames(a4, "RT", "Clitocine")
  a5<-treated_samples[treatment=="Gentamicin" & replicate==1, c("identifier","RT")]; setnames(a5, "RT", "Gentamicin")
  a6<-treated_samples[treatment=="G418" & replicate==1, c("identifier","RT")]; setnames(a6, "RT", "G418")
  a7<-treated_samples[treatment=="CC90009" & replicate==1, c("identifier","RT")]; setnames(a7, "RT", "CC90009")
  a8<-treated_samples[treatment=="SJ6986" & replicate==1, c("identifier","RT")]; setnames(a8, "RT", "SJ6986")
  a9<-treated_samples[treatment=="FUr" & replicate==1, c("identifier","RT")]; setnames(a9, "RT", "FUr")
```  
Join them in 'all_drugs_RT_ascolumns' object, to plot Fig.2c scatterplots. 'all_drugs_RT_ascolumns' has the drug-specific RT as columns together with the stop type and downstream sequence context information.

```
  all_drugs_RT_ascolumns<-left_join(left_join(left_join(left_join(left_join(left_join(left_join(left_join(a1, a2, by="identifier"), a3,by="identifier"), a4, by="identifier"), a5,by="identifier"), a6, by="identifier"), a7,by="identifier"),a8,by="identifier"),a9,by="identifier")
```

Alternatively, directly load the 'all_drugs_RT_ascolumns.rds' object from base_dir:

```
all_drugs_RT_ascolumns<-readRDS(file = file.path(base_dir, "Fig2/all_drugs_RT_ascolumns.rds"))

```

SJ6986 vs SRI scatterplot:

```
        ggplot(all_drugs_RT_ascolumns) + geom_hex(aes(y=SRI, x=SJ6986, fill=toupper(stop_type)), bins=70) + stops_fills + labs(y="SRI", x="SJ6986") + stat_cor(aes(y=SRI, x=SJ6986, label = ..r.label..), size=10, label.y.npc=c(1,1)) +
          p + theme(panel.spacing.y = unit(.5, "lines"), legend.position = "right", text = element_text(size = 25)) 
```        
Clitocine vs SRI scatterplot:

```
        ggplot(all_drugs_RT_ascolumns) + geom_hex(aes(y=SRI, x=Clitocine, fill=toupper(stop_type)), bins=70) + stops_fills + labs(y="Clitocine", x="SRI") + stat_cor(aes(y=SRI, x=Clitocine, label = ..r.label..), size=10, label.y.npc=c(1,1)) +
          p + theme(panel.spacing.y = unit(.5, "lines"), legend.position = "right", text = element_text(size = 25)) 
```

**Fig.2c:** Readthrough heatmap.

Generate heatmap_df, which is a reduction of the treated_samples: a) select only the treatment, identifier and RT columns. b) only one replicate and c) remove all the variants with at least one NA value in any of the treatments (n=300). The clustering doesn't know how to deal with NAs. Also, if one treatment has NA value, it means that these variants was in very low frequency in the library and probably has v low coverage & high error.

```
    heatmap_df<-treated_samples[replicate==1, c("treatment", "identifier", "RT")]
    identifiers_to_remove<-unique(heatmap_df[is.na(RT), identifier])
    heatmap_df<-heatmap_df[!identifier%chin%identifiers_to_remove]
  ```
  'r' is a rearrangement of heatmap_df, to do the clustering. Basically each variant becomes a column (ncols~5500), and the treatments are the rows. Then, I do the clustering on columns and store it on the 'clust' object. In the ggplot I'm gonna use the clust object to sort the y axis.
    
```
    r<-tidyr::pivot_wider(heatmap_df, names_from = "identifier", values_from = "RT")
    r <- as.matrix(r[, -1]) # remove the treatment column to do the clustering (we don't want it, it would mess things up)
    clust_rows_heatmap <- hclust(dist(t(r)))
```
  Do the clustering for the columns (We want the heatmap to have both axis clustered)
```
    c<-tidyr::pivot_wider(heatmap_df, names_from = "treatment", values_from = "RT")
    c <- as.matrix(c[, -1]) # remove the variant column to do the clustering (we don't want it, it would mess things up)
    clust_cols_heatmap <- hclust(dist(t(c)))
```

Alternatively, directly load the 'heatmap_df.rds' object from base_dir:

```
  heatmap_df<-readRDS(file = file.path(base_dir, "Fig2/heatmap_df.rds"))

```

  Plot:
```
    plot<-ggplot(heatmap_df, aes(x=treatment,y=as.character(identifier))) + geom_tile(aes(fill=RT))+scale_fill_gradient2(low="red",mid="white",high="blue",midpoint=0, limits = c(0, 7), oob = scales::squish) + 
      scale_y_discrete(limits = colnames(r)[clust_rows_heatmap$order]) + scale_x_discrete(limits=colnames(c)[clust_cols_heatmap$order]) + labs(x="Treatment", y="Variants", fill="RT(%)") + p +theme(axis.text.y = element_blank(), axis.text.x = element_text(size=12.5))
```
  
  
**Fig.2d:**Stop type
```
ggplot(treated_samples[treatment%in%c("Clitocine", "SRI", "DAP", "G418") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(stop_type), fill=treatment)) + ylim(0,4) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="Stop type", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = c(.5,.95), legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))

```

**Fig.2e**: +1 nt downstream of the TGA PTC.
```        
          ggplot(treated_samples[stop_type=="tga" & treatment%in%c("Clitocine", "SRI", "DAP", "G418") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(down_1nt), fill=treatment)) + ylim(0,5.5) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="+1nt downstream the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = "none", legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```

**Fig.2f:** Stop type and +1nt downstream interaction.

```
ggplot(treated_samples[treatment%in%c("Clitocine", "SRI") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(down_1nt), fill=treatment)) + geom_boxplot(outlier.shape = NA) + drugs_fills + labs(x="Stoptype*down_1nt interaction", y="Readthrough (%)", title="") + 
            theme(legend.text = element_text(size=12), legend.title=element_blank(),legend.position = c(.5,1.15), legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55)) +
            facet_wrap(~toupper(stop_type), scales = "free") + p
```

**Fig.2g:**Trinucleotide downstream of the TGA PTC for SRI drug.
```
ggplot(treated_samples[stop_type=="tga" & treatment%in%c("SRI") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(down_123nt), fill=treatment)) + ylim(0,6.5) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="+123nt downstream the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), axis.text.x=element_text(angle=90, size=10), legend.title = element_text(size=15), legend.position = c(.8,.95), legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```

**Fig.2h**: Trinucleotide upstream of the TGA PTC for SRI drug.
```
ggplot(treated_samples[stop_type=="tga" & treatment%in%c("SRI") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(up_123nt), fill=treatment)) + ylim(0,6.5) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="+123nt upstream the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), axis.text.x=element_text(angle=90, size=10), legend.title = element_text(size=15), legend.position = c(.8,.95), legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```

**Fig.2i:**Glutamic acid upstream of the TGA PTC.
```
ggplot(treated_samples[stop_type=="tga" & up_aa=="E" & treatment%in%c("Clitocine", "SRI", "DAP", "G418") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(up_123nt), fill=treatment)) + ylim(0,4.5) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="Glutamic acid upstream of the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = "none", legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```
 
**Fig.2j:**

```
ggplot(treated_samples[stop_type=="tga" & treatment%in%c("Clitocine", "SRI", "DAP", "G418") & up_aa%chin%c("E", "G", "L", "K", "Q", "V", "S", "I", "R", "P", "T", "A") & replicate==1 & reads_allbins>20], aes(y=RT, x=up_aa, fill=toupper(up_1nt))) + geom_boxplot(outlier.shape = NA) + 
            nts_fills + labs(x="Aminoacid upstream of the TGA PTC", y="Readthrough (%)", fill="3rd nt") + p + facet_wrap(~treatment, nrow=2, scales = "free") +
            theme(panel.spacing.y = unit(.5, "lines"), legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = "top", legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```            
            
            
**Extended Data Fig.2:**

**Ext.Data.Fig.2a:**

```  
ggplot(treated_samples[stop_type=="tga" & treatment%in%c("Clitocine", "SRI", "DAP", "G418") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(down_123nt), fill=treatment)) + ylim(0,6.5) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="Three nucleotides downstream the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), axis.text.x=element_text(angle=90, size=10), legend.title = element_text(size=15), legend.position = c(.8,.95), legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```          

**Ext.Data.Fig.2b:**

``` 
ggplot(treated_samples[stop_type=="tga" & treatment%in%c("Clitocine", "SRI", "DAP", "G418") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(up_123nt), fill=treatment)) + ylim(0,6.5) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="Three nucleotides upstream the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), axis.text.x=element_text(angle=90, size=10), legend.title = element_text(size=15), legend.position = c(.5,.95), legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```          

**Ext.Data.Fig.2c:**
```
ggplot(treated_samples[treatment%in%c("SJ6986", "CC90009", "FUr", "Gentamicin") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(stop_type), fill=treatment)) + ylim(0,4) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="Stop type", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = c(.5,.95), legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```
**Ext.Data.Fig.2d:**
```
          ggplot(treated_samples[stop_type=="tga" & treatment%in%c("SJ6986", "CC90009", "FUr", "Gentamicin") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(down_1nt), fill=treatment)) + ylim(0,5.5) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="+1nt downstream the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = "none", legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```    

**Ext.Data.Fig.2e:**

```
ggplot(treated_samples[treatment%in%c("G418", "SJ6986", "DAP") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(down_1nt), fill=treatment)) + geom_boxplot(outlier.shape = NA) + drugs_fills + labs(x="Stoptype*down_1nt interaction", y="Readthrough (%)", title="") + 
            theme(legend.text = element_text(size=12), legend.title=element_blank(),legend.position = c(.5,1.15), legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55)) +
            facet_wrap(~toupper(stop_type), scales = "free") + p
```

**Ext.Data.Fig.2f:**

```
ggplot(treated_samples[stop_type=="tga" & treatment%in%c("Clitocine", "SRI", "DAP", "G418") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(up_aa), fill=treatment)) + ylim(0,6) + geom_boxplot(outlier.shape = NA) + drugs_fills + labs(x="Aminoacid upstream of the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
    theme(legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = c(.5,.95), legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```

**Ext.Data.Fig.2g:**

```
ggplot(treated_samples[stop_type=="tga" & up_aa=="E" & treatment%in%c("SJ6986", "CC90009", "FUr", "Gentamicin") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(up_123nt), fill=treatment)) + ylim(0,4.5) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="Glutamic acid upstream of the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = "none", legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```

**Ext.Data.Fig.2h:**

```
ggplot(treated_samples[stop_type=="tga" & up_aa=="R" & treatment%in%c("Clitocine", "SRI", "DAP", "G418") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(up_123nt), fill=treatment)) + ylim(0,4.5) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="Arginine upstream of the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = "none", legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```

**Ext.Data.Fig.2i:**

```
ggplot(treated_samples[stop_type=="tga" & up_aa=="R" & treatment%in%c("SJ6986", "CC90009", "FUr", "Gentamicin") & replicate==1 & reads_allbins>20], aes(y=RT, x=toupper(up_123nt), fill=treatment)) + ylim(0,4.5) + geom_boxplot(outlier.shape = NA) + 
            drugs_fills + labs(x="Arginine upstream of the TGA PTC", y="Readthrough (%)", title="", fill="Treatment") + p + 
            theme(legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = "none", legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```

**Ext.Data.Fig.2j:**

```
 ggplot(treated_samples[stop_type=="tga" & treatment%in%c("SJ6986", "CC90009") & up_aa%chin%c("E", "G", "L", "K", "Q", "V", "S", "I", "R", "P", "T", "A") & replicate==1 & reads_allbins>20], aes(y=RT, x=up_aa, fill=toupper(up_1nt))) + geom_boxplot(outlier.shape = NA) + 
            nts_fills + labs(x="Aminoacid upstream of the TGA PTC", y="Readthrough (%)", fill="3rd nt") + p + facet_wrap(~treatment, nrow=2, scales = "free") +
            theme(panel.spacing.y = unit(.5, "lines"), legend.text = element_text(size=12), legend.title = element_text(size=15), legend.position = "top", legend.direction="horizontal") + guides(fill = guide_legend(title.position = "top", title.hjust=.55))
```

**Ext.Data.Fig.2k:**

Calculate the mean difference in readthrough between 1 HDist codons in same_aa vs different_aa -> to test if there is an aminoacid-driven effect of the 3nts upstream.

The calculations are saved in the 'one_hdist_codons_RTcomparision' object:

```
  drugs<-c("FUr", "Gentamicin", "CC90009", "G418", "Clitocine", "DAP", "SJ6986", "SRI")
  one_hdist_codons_RTcomparision<-data.frame(same_aa=0, different_aa=0, pvalue=0, RTCs="tofill")
  for (j in 1:length(drugs)){
    tmp<-treated_samples[stop_type=="tga" & treatment==drugs[j] & replicate==1, mean(RT, na.rm=T), by=up_123nt]
    a1<-as.data.table(t(combn(tmp$up_123nt,2)))
    setnames(a1, c("V1", "V2"), c("codon_1", "codon_2"))
    for (i in 1:nrow(a1)){
      a1$RT_1[i]<-tmp[up_123nt==a1$codon_1[i], V1]
      a1$RT_2[i]<-tmp[up_123nt==a1$codon_2[i], V1]
    }
    a1[, aa_1:=translate(s2c(codon_1)), by=1:nrow(a1)]
    a1[, aa_2:=translate(s2c(codon_2)), by=1:nrow(a1)]
    a1[, same_aa:=if_else(aa_1==aa_2, "yes", "no")]
    a1[, RT_diff:=abs(RT_1-RT_2)]
    a1[, hdist:=sum(s2c(codon_1)!=s2c(codon_2)), by=1:nrow(a1)]
    a1<-a1[hdist==1]
    group1<-a1[same_aa=="yes", RT_diff]
    group2<-a1[same_aa=="no", RT_diff]
    one_hdist_codons_RTcomparision[j,1]<-round(mean(group1),4)
    one_hdist_codons_RTcomparision[j,2]<-round(mean(group2),4)
    one_hdist_codons_RTcomparision[j,3]<-round(t.test(group1, group2)$p.value,4)
    one_hdist_codons_RTcomparision[j,4]<-drugs[j]
  }
  one_hdist_codons_RTcomparision<-as.data.table(one_hdist_codons_RTcomparision)
  one_hdist_codons_RTcomparision
```