---
title: "Fig3_extdataFig3"
output: html_document
---

**Load the 'treated_samples.rds' object:**

``` 
  treated_samples<-readRDS(file = "/base_dir/treated_samples.rds")
```

**Fig.3a:** Readthrough efficiencies for all PTEN mutations measured in our dataset. Each drug-specific plot is stored in p1,p2,...,p9.

```
p1<-ggplot(treated_samples[treatment=="Clitocine" & replicate==1 & GENEINFO=="TP53"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="Clitocine" & replicate==1 & GENEINFO=="TP53"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="Clitocine") + p + theme(axis.text.x = element_blank()) + scale_y_continuous(breaks=seq(1, 3, 1))  +
            geom_point(data=treated_samples[treatment=="Clitocine" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="Clitocine" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], nudge_y = 0.25, aes(label=mutation_identifier_short), color="black")
          
          p2<-ggplot(treated_samples[treatment=="SRI" & replicate==1 & GENEINFO=="TP53"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="SRI" & replicate==1 & GENEINFO=="TP53"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="SRI") + p + theme(axis.text.x = element_blank()) +
            geom_point(data=treated_samples[treatment=="SRI" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="SRI" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], nudge_y = 1, aes(label= mutation_identifier_short), color="black")
          
          p3<-ggplot(treated_samples[treatment=="G418" & replicate==1 & GENEINFO=="TP53"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="G418" & replicate==1 & GENEINFO=="TP53"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="G418") + p + theme(axis.text.x = element_blank()) +
            geom_point(data=treated_samples[treatment=="G418" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="G418" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], position=position_jitter(height =.75), aes(label= mutation_identifier_short), color="black")
          
          p4<-ggplot(treated_samples[treatment=="DAP" & replicate==1 & GENEINFO=="TP53"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="DAP" & replicate==1 & GENEINFO=="TP53"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="DAP") + p + theme(axis.text.x = element_blank()) + 
            geom_point(data=treated_samples[treatment=="DAP" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="DAP" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], nudge_y = 0.5, aes(label=mutation_identifier_short), color="black")
          
          p5<-ggplot(treated_samples[treatment=="SJ6986" & replicate==1 & GENEINFO=="TP53"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="SJ6986" & replicate==1 & GENEINFO=="TP53"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="SJ6986") + p + theme(axis.text.x = element_blank()) + 
            geom_point(data=treated_samples[treatment=="SJ6986" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="SJ6986" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], nudge_y = 0.5, aes(label=mutation_identifier_short), color="black")
          
          p6<-ggplot(treated_samples[treatment=="CC90009" & replicate==1 & GENEINFO=="TP53"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="CC90009" & replicate==1 & GENEINFO=="TP53"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="CC90009") + p + theme(axis.text.x = element_blank()) + scale_y_continuous(breaks=seq(1,2, 1))  + 
            geom_point(data=treated_samples[treatment=="CC90009" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="CC90009" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")],  position=position_jitter(height =.3), aes(label=mutation_identifier_short), color="black")
          
          p7<-ggplot(treated_samples[treatment=="FUr" & replicate==1 & GENEINFO=="TP53"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() + 
            scale_x_discrete(limits = sort(treated_samples[treatment=="FUr" & replicate==1 & GENEINFO=="TP53"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="FUr") + p + theme(axis.text.x = element_blank())  + 
            geom_point(data=treated_samples[treatment=="FUr" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="FUr" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X")], nudge_y = 0.25, aes(label=mutation_identifier_short), color="black")
          
          p8<-ggplot(treated_samples[treatment=="Gentamicin" & replicate==1 & GENEINFO=="TP53"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() + 
            scale_x_discrete(limits = sort(treated_samples[treatment=="Gentamicin" & replicate==1 & GENEINFO=="TP53"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="Gentamicin") + p + theme(axis.text.x = element_blank())  + 
            geom_point(data=treated_samples[treatment=="Gentamicin" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="Gentamicin" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X")],  nudge_y = 0.15, aes(label=mutation_identifier_short), color="black")
          
          p9<-ggplot(treated_samples[treatment=="Untreated" & replicate==1 & GENEINFO=="TP53"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() + 
            scale_x_discrete(limits = sort(treated_samples[treatment=="Untreated" & replicate==1 & GENEINFO=="TP53"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="Untreated") + p + theme(axis.text.x = element_blank()) + scale_y_continuous(breaks=seq(0.1, 1, 0.2))  + 
            geom_point(data=treated_samples[treatment=="Untreated" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X", "R196X", "Q192X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="Untreated" & replicate==1 & GENEINFO=="TP53" &  mutation_identifier_short%chin%c("R213X", "R342X", "R306X")], nudge_y = 0.075, aes(label=mutation_identifier_short), color="black")
```          
          
Gather the drug individual plots in groups of three:
          
```
          p10<-ggarrange(p1 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()), 
                         p2 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                         p3 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                         ncol=1, nrow=3, common.legend = TRUE, legend="bottom")
          annotate_figure(p10, left = textGrob("Readthrough (%)", rot = 90, vjust = .5, gp = gpar(fontsize = 20)),
                          bottom = textGrob("*TP53* transcript (5'-3')", gp = gpar(fontsize = 20)))
          p11<-ggarrange(p4 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()), 
                         p5 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                         p6 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                         ncol=1, nrow=3, common.legend = TRUE, legend="bottom")
          annotate_figure(p11, left = textGrob("Readthrough (%)", rot = 90, vjust = .5, gp = gpar(fontsize = 20)),
                          bottom = textGrob("*TP53* transcript (5'-3')", gp = gpar(fontsize = 20)))
          p12<-ggarrange(p7 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()), 
                         p8 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                         p9 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                         ncol=1, nrow=3, common.legend = TRUE, legend="bottom")
          annotate_figure(p12, left = textGrob("Readthrough (%)", rot = 90, vjust = .5, gp = gpar(fontsize = 20)),
                          bottom = textGrob("*TP53* transcript (5'-3')", gp = gpar(fontsize = 20)))
          
```


**Fig.3b:** Overlap of the top50 RT variants for each treatment:

Select the top50 variants, and annotate the overlap between treatments (pair-wise) in df object:

```
  tmp<-as.data.table(treated_samples[replicate==1 & viral=="no", c("RT", "treatment", "mutation_identifier")] %>% dplyr::group_by(treatment) %>% dplyr::top_n(50, RT) %>% ungroup())
  drugs<-c("Untreated", "FUr", "Gentamicin", "CC90009", "G418", "Clitocine", "DAP", "SJ6986", "SRI")
  vec1<-c(); vec2<-c(); vec3<-c()
  for (i in 1:length(drugs)){
    for (j in 1:length(drugs)){
      vec1<-c(vec1,length(which(tmp[treatment==drugs[i], mutation_identifier]%chin%tmp[treatment==drugs[j], mutation_identifier])))
      vec2<-c(vec2, drugs[i])
      vec3<-c(vec3, drugs[j])
    }}
  df<-as.data.table(data.frame(treatment_1=vec2, treatment_2=vec3, intersection=vec1))
  
```  
Convert from wide to long to have a correlation matrix:
```
  df<-spread(df, treatment_2, intersection)
  # Generate a function that eliminates the redudant data of the correlation matrix!
  get_upper_tri <- function(df){
    df[upper.tri(df)]<- NA
    return(df)
  }
  upper_tri <- get_upper_tri(df)
  top50vars_byRTC<- melt(upper_tri, na.rm = TRUE)
  top50vars_byRTC<-as.data.table(top50vars_byRTC)
```  

Alternatively, directly load the 'top50vars_byRTC.rds' object from base_dir:

```
  top50vars_byRTC<-readRDS(file = "/base_dir/Fig3/top50vars_byRTC.rds")
```

Plot the heatmap:
```
  ggplot(top50vars_byRTC, aes(y=variable ,x=treatment_1, fill=value, label=value)) + geom_tile()+scale_fill_gradient2(low="red",mid="white",high="blue",midpoint=0, limits = c(0, 50), oob = scales::squish) + 
    geom_text() + labs(fill="#Overlap top50 PTCs", x="", y="") + p 

```


**Fig.3c:**

Objects needed (generated for Fig.2):

```
  all_drugs_RT_ascolumns<-readRDS(file = "/base_dir/Fig2/all_drugs_RT_ascolumns.rds")
```

We use all_drugs_RT_ascolumns to generate 'fig3_thresholds'. This dtbl contains the percentage of mutations (out of the 5837) with RT>treshold (tresholds range from 1-7) for each drug and considering all drugs together:
  In this loop we calculate it for each drug separately:
```
    tmp1<-as.data.table(gather(all_drugs_RT_ascolumns, Treatment, RT, c("Untreated", "SRI", "DAP", "Clitocine", "Gentamicin", "G418", "CC90009", "SJ6986", "FUr")))
    for (i in 1:6){
      if (i==1){
        fig3_thresholds<-tmp1[RT>i, (.N/5837)*100, by=Treatment][, Threshold:=i]
      } else {
        fig3_thresholds<-rbind(fig3_thresholds, tmp1[RT>i, (.N/5837)*100, by=Treatment][, Threshold:=i])
      }}
    fig3_thresholds$Threshold<-as.factor(fig3_thresholds$Threshold)
    setnames(fig3_thresholds, c("V1", "Threshold"), c("% Mutations", "Threshold(%)"))
```
  This loop is to compute the percentage of mutations (out of the 5837) with RT>treshold when considering any of the eight drugs:
  
```
    tmp2<-as.data.frame(data.table(Treatment="All_drugs", "% Mutations"=100, "Threshold(%)"=10))
    for (i in 1:6){
      tmp2[i,1]<-"All_drugs"
      tmp2[i,2]<-(all_drugs_RT_ascolumns[CC90009>i | G418>i | Clitocine>i | DAP>i | SJ6986>i | SRI>i | FUr>i | Gentamicin>i , .N]/nrow(all_drugs_RT_ascolumns))*100
      tmp2[i,3]<-i
    }
```
  Now we join the tmp2 with the fig3_thresholds, to have the complete fig3_thresholds object: 
```  
    fig3_thresholds<-rbind(fig3_thresholds, tmp2)
```

Alternatively, directly load the 'fig3_thresholds.rds' object from base_dir:

```
  fig3_thresholds<-readRDS(file = "/base_dir/Fig3/fig3_thresholds.rds")
```

Plot:
```
ggplot(fig3_thresholds[!Treatment%chin%c("Untreated")], aes(x=Treatment, y=`% Mutations`, fill=`Threshold(%)`)) + geom_bar(stat="identity", position = "dodge") + 
          p + thresholds_fills + theme(axis.text.x = element_text(size=15, angle=90, hjust=1, vjust=0.5), axis.title.x = element_blank(), legend.position = "top")
      
```



**Fig.3d:** Multistop variants.
Generate the multistop_3 datatable. THe steps are the following:
Identify the multistop variants in one of the treatments (here using Clitocine):
```
  multistops_1<-treated_samples[treatment=="Clitocine" & replicate==1 & multistop=="yes" & reads_allbins>15, ]
  ```
Because of the 15reads filter above, some variant of the mulitstop pair might pass the filter but the other doesn't. We have to get rid of those because they are not longer useful:

```
  retained<-multistops_1[duplicated(aa_seq), aa_seq]
  multistops_1<-multistops_1[aa_seq%chin%retained,]
```
Also, there are some variants that are mislabeled as multistop -> they have the same aa_seq but different nt_seq (one point mutation not in the PTC that results in the same aa_seq), and because of how I designed the library, these were mislabeled as multistop. Eliminate them:

```
  aa<-multistops_1[,unique(stop_type),by=aa_seq]
  aa<-aa[which(duplicated(aa$aa_seq)==T), aa_seq]
  retained_2<-multistops_1[aa_seq%chin%aa, aa_seq]
  multistops_1<-multistops_1[aa_seq%chin%retained_2]
```

Now retrieve the selected multistop sequences from the other treatments, and merge all treatments in multistop_2 df:
```
  multistops_2<-rbind(rbind(rbind(rbind(rbind(rbind(rbind(multistops_1, treated_samples[treatment=="SRI" & replicate==1 & aa_seq%chin%multistops_1$aa_seq]), treated_samples[treatment=="DAP" & replicate==1 & aa_seq%chin%multistops_1$aa_seq]), treated_samples[treatment=="SJ6986" & replicate==1 & aa_seq%chin%multistops_1$aa_seq]), treated_samples[treatment=="G418" & replicate==1 & aa_seq%chin%multistops_1$aa_seq]), treated_samples[treatment=="CC90009" & replicate==1 & aa_seq%chin%multistops_1$aa_seq]), treated_samples[treatment=="FUr" & replicate==1 & aa_seq%chin%multistops_1$aa_seq]), treated_samples[treatment=="Gentamicin" & replicate==1 & aa_seq%chin%multistops_1$aa_seq])
  toeliminate<-multistops_2[(treatment=="SRI" & reads_allbins<15) | (treatment=="DAP" & reads_allbins<15), unique(aa_seq)]
  multistops_2<-multistops_2[!(aa_seq%chin%toeliminate)]
```

Alternatively, directly load the 'multistops_2.rds' object from base_dir:

```
  multistops_2<-readRDS(file = "/base_dir/Fig3/multistops_2.rds")
```


All sequences can't be plotted together, because the plot would be unreadable. We have two options:

  A. Randomly select a number of them (i.e.12).
  
```
  variants<-sample(unique(multistops_2$aa_seq),12)
```
  B. Targeted selection.
  
```
  variants<-unique(multistops_2[mutation_identifier%chin%c("BRCA2_Y2977X", "APC_Y1376X", "IFNGR1_S306X", "BRCA1_S1207X", "BRCA2_S2835X", "NF1_S365X", "FLCN_Y463X", "PTEN_Y88X", "APC_S583X", "PMS2_S459X", "PALB2_Y1055X", "DMD_S622X"), aa_seq])
```
  
  Store the variants to be plotted in multistops_3:
  
```
  multistops_3<-multistops_2[aa_seq%chin%variants]
  multistops_3<-multistops_3[treatment%in%c("SRI", "Clitocine", "DAP")]

```

Plot:

```
  ggplot(multistops_3[treatment%in%c("SRI", "G418", "Clitocine", "DAP")], aes(x = interaction(as.character(mutation_identifier), as.character(treatment), lex.order = TRUE), y=RT, fill=toupper(stop_type))) +
            geom_bar(stat="identity", position="stack", color="black") +
            stops_fills +
            annotate(geom = "text", x = seq_len((nrow(multistops_3)/2)), y = -0.4, label = rep(c("Clito", "DAP", "G418", "SRI"),nrow(multistops_3)/8), size = 3, angle=90) +
            annotate(geom = "text", x = 2.5 + 4 * (0:((nrow(multistops_3)/8)-1)), y = -.95, label = sort(as.character(unique(multistops_3$mutation_identifier))), size = 2.7) +
            annotate("segment", x = seq(0.8, nrow(multistops_3)/2 ,4), xend = seq(4.2, (nrow(multistops_3)/2)+1,4), y = -0.75, yend = -0.75) +
            coord_cartesian(ylim = c(0, 10), expand = FALSE, clip = "off") +
            theme_bw() +
            geom_vline(xintercept = 4.5+4*(1:((nrow(multistops_3)-1)/8)-1), linetype="dotted", 
                       color = "blue", size=1) + labs(fill='Stop Type', y="Readthrough (%)") +
            theme(plot.margin = unit(c(1, 1, 4, 1), "lines"),
                  axis.title.x = element_blank(),
                  axis.text.x = element_blank(),
                  legend.position = c(0.9, .85),
                  legend.text = element_text(size=15)) + p
          
```












**Ext.Data.Fig.3a:**
```
  ggplot(treated_samples) + geom_hex(aes(x=RT, y=tAI_up_5aas), bins=70) + geom_smooth(aes(x=RT, y=tAI_up_5aas), method='lm') + stat_cor(aes(x=RT, y=tAI_up_5aas, label = ..r.label..), size=4, label.y.npc=c(1,1)) + facet_wrap(~treatment, nrow=3, scales = "free") + labs(y="tAI 5 codons upstream", x="Readthrough (%)") +
            scale_fill_viridis(limits = c(0, 100), oob = scales::squish, alpha = 0.85) + p
```

**Ext.Data.Fig.3b:**

```
        ggplot(treated_samples) + geom_hex(aes(x=RT, y=CAI_up_5aas), bins=70) + geom_smooth(aes(x=RT, y=CAI_up_5aas), method='lm') + stat_cor(aes(x=RT, y=CAI_up_5aas, label = ..r.label..), size=4, label.y.npc=c(1,1)) + facet_wrap(~treatment, nrow=3, scales = "free") + labs(y="CAI 5 codons upstream", x="Readthrough (%)") +
            scale_fill_viridis(limits = c(0, 100), oob = scales::squish, alpha = 0.85) + p
```

**Ext.Data.Fig.3c:**

```               
        ggplot(treated_samples) + geom_point(aes(x=RT, y=GCperc_5aas), alpha=.5, color="#FFCCCC") + geom_smooth(aes(x=RT, y=GCperc_5aas), method='lm') + stat_cor(aes(x=RT, y=GCperc_5aas, label = ..r.label..), size=4, label.y.npc=c(1,1)) + facet_wrap(~treatment, nrow=3, scales = "free") + labs(y="GC% 5 codons upstream", x="Readthrough (%)") +
            scale_fill_viridis(limits = c(0, 100), oob = scales::squish, alpha = 0.85) + p
```


**Ext.Data.Fig.3d:** readthrough efficiencies for all PTEN mutations measured in our dataset. Each drug-specific plot is stored in p1,p2,...,p9.

```
    p1<-ggplot(treated_samples[treatment=="Clitocine" & replicate==1 & GENEINFO=="PTEN"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="Clitocine" & replicate==1 & GENEINFO=="PTEN"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="Clitocine") + p + theme(axis.text.x = element_blank()) + scale_y_continuous(breaks=seq(1, 2, 1))  + 
            geom_point(data=treated_samples[treatment=="Clitocine" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="Clitocine" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], nudge_y = 0.25, aes(label=mutation_identifier_short), color="black")
            
    p2<-ggplot(treated_samples[treatment=="SRI" & replicate==1 & GENEINFO=="PTEN"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="SRI" & replicate==1 & GENEINFO=="PTEN"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="SRI") + p + theme(axis.text.x = element_blank()) +
            geom_point(data=treated_samples[treatment=="SRI" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="SRI" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], nudge_y = 0.75, aes(label= mutation_identifier_short), color="black")
          
    p3<-ggplot(treated_samples[treatment=="G418" & replicate==1 & GENEINFO=="PTEN"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="G418" & replicate==1 & GENEINFO=="PTEN"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="G418") + p + theme(axis.text.x = element_blank()) +
            geom_point(data=treated_samples[treatment=="G418" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="G418" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], position=position_jitter(height =.25), aes(label= mutation_identifier_short), color="black")

    p4<-ggplot(treated_samples[treatment=="DAP" & replicate==1 & GENEINFO=="PTEN"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="DAP" & replicate==1 & GENEINFO=="PTEN"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="DAP") + p + theme(axis.text.x = element_blank()) + 
            geom_point(data=treated_samples[treatment=="DAP" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="DAP" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], nudge_y = 0.5, aes(label=mutation_identifier_short), color="black")
          
    p5<-ggplot(treated_samples[treatment=="SJ6986" & replicate==1 & GENEINFO=="PTEN"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="SJ6986" & replicate==1 & GENEINFO=="PTEN"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="SJ6986") + p + theme(axis.text.x = element_blank()) + 
            geom_point(data=treated_samples[treatment=="SJ6986" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="SJ6986" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], nudge_y = 0.5, aes(label=mutation_identifier_short), color="black")
          
    p6<-ggplot(treated_samples[treatment=="CC90009" & replicate==1 & GENEINFO=="PTEN"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() +
            scale_x_discrete(limits = sort(treated_samples[treatment=="CC90009" & replicate==1 & GENEINFO=="PTEN"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="CC90009") + p + theme(axis.text.x = element_blank()) + scale_y_continuous(breaks=seq(1,2,1))  + 
            geom_point(data=treated_samples[treatment=="CC90009" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="CC90009" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")],  position=position_jitter(height =.2), aes(label=mutation_identifier_short), color="black")
          
    p7<-ggplot(treated_samples[treatment=="FUr" & replicate==1 & GENEINFO=="PTEN"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() + 
            scale_x_discrete(limits = sort(treated_samples[treatment=="FUr" & replicate==1 & GENEINFO=="PTEN"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="FUr") + p + theme(axis.text.x = element_blank())  + 
            geom_point(data=treated_samples[treatment=="FUr" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="FUr" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")],  position=position_jitter(height =.1), aes(label=mutation_identifier_short), color="black")
          
    p8<-ggplot(treated_samples[treatment=="Gentamicin" & replicate==1 & GENEINFO=="PTEN"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() + 
            scale_x_discrete(limits = sort(treated_samples[treatment=="Gentamicin" & replicate==1 & GENEINFO=="PTEN"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="Gentamicin") + p + theme(axis.text.x = element_blank())  + 
            geom_point(data=treated_samples[treatment=="Gentamicin" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="Gentamicin" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")],  position=position_jitter(height =.05), aes(label=mutation_identifier_short), color="black")
          
    p9<-ggplot(treated_samples[treatment=="Untreated" & replicate==1 & GENEINFO=="PTEN"], aes(x=aa_gene_position, y=RT, group=toupper(stop_type), color=toupper(stop_type)))+ geom_point() + 
            scale_x_discrete(limits = sort(treated_samples[treatment=="Untreated" & replicate==1 & GENEINFO=="PTEN"]$aa_gene_position)) + stops_colors + 
            labs(y="Readthrough (%)", x="Transcript position (5'-3')", title="Untreated") + p + theme(axis.text.x = element_blank()) + scale_y_continuous(breaks=seq(0.1, 0.4, 0.1))  + 
            geom_point(data=treated_samples[treatment=="Untreated" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")], size=4, fill="black") +
            geom_text(data=treated_samples[treatment=="Untreated" & replicate==1 & GENEINFO=="PTEN" &  mutation_identifier_short%chin%c("R130X", "R233X", "Q171X", "Q245X")],  position=position_jitter(height =.025), aes(label=mutation_identifier_short), color="black")
```          
          
Gather the drug individual plots in groups of three: 

```
        p10<-ggarrange(p1 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()), 
                       p2 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                       p3 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                       ncol=1, nrow=3, common.legend = TRUE, legend="bottom")
        annotate_figure(p10, left = textGrob("Readthrough (%)", rot = 90, vjust = .5, gp = gpar(fontsize = 20)),
                        bottom = textGrob("*PTEN* transcript (5'-3')", gp = gpar(fontsize = 20)))
        p11<-ggarrange(p4 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()), 
                         p5 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                         p6 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                         ncol=1, nrow=3, common.legend = TRUE, legend="bottom")
          annotate_figure(p11, left = textGrob("Readthrough (%)", rot = 90, vjust = .5, gp = gpar(fontsize = 20)),
                          bottom = textGrob("*PTEN* transcript (5'-3')", gp = gpar(fontsize = 20)))
        p12<-ggarrange(p7 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()), 
                         p8 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                         p9 + theme(axis.title.x = element_blank(), axis.title.y = element_blank()),
                         ncol=1, nrow=3, common.legend = TRUE, legend="bottom")
          annotate_figure(p12, left = textGrob("Readthrough (%)", rot = 90, vjust = .5, gp = gpar(fontsize = 20)),
                          bottom = textGrob("*PTEN* transcript (5'-3')", gp = gpar(fontsize = 20)))
          
 ``` 




**Ext.Data.Fig.3e: **Correlation between stop types for each drug:
Create the plots for each drug and stop type pairwise combination separately (p1, p2, p3, ..., p16). Then combine them in p17a, p17b; and finally in p17.

```
  tga_stops<-multistops_2[treatment=="DAP" &stop_type=="tga", c("start", "RT")]; setnames(tga_stops, "RT", "RT_TGA")
  taa_stops<-multistops_2[treatment=="DAP" &stop_type=="taa", c("start", "RT")]; setnames(taa_stops, "RT", "RT_TAA")
  tag_stops<-multistops_2[treatment=="DAP" &stop_type=="tag", c("start", "RT")]; setnames(tag_stops, "RT", "RT_TAG")
  tga_taa_stops<-full_join(full_join(tga_stops,taa_stops, by="start"), tag_stops, by="start")
  p1<-ggplot(tga_taa_stops[!is.na(RT_TAG)], aes(x=RT_TAA, y=RT_TAG)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("DAP") + p
  p2<-ggplot(tga_taa_stops[!is.na(RT_TGA)], aes(x=RT_TAA, y=RT_TGA)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("DAP")  + p
  
  tga_stops<-multistops_2[treatment=="SRI" &stop_type=="tga", c("start", "RT")]; setnames(tga_stops, "RT", "RT_TGA")
  taa_stops<-multistops_2[treatment=="SRI" &stop_type=="taa", c("start", "RT")]; setnames(taa_stops, "RT", "RT_TAA")
  tag_stops<-multistops_2[treatment=="SRI" &stop_type=="tag", c("start", "RT")]; setnames(tag_stops, "RT", "RT_TAG")
  tga_taa_stops<-full_join(full_join(tga_stops,taa_stops, by="start"), tag_stops, by="start")
  p3<-ggplot(tga_taa_stops[!is.na(RT_TAG)], aes(x=RT_TAA, y=RT_TAG)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("SRI") + p
  p4<-ggplot(tga_taa_stops[!is.na(RT_TGA)], aes(x=RT_TAA, y=RT_TGA)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("SRI")  + p
  
  tga_stops<-multistops_2[treatment=="Clitocine" &stop_type=="tga", c("start", "RT")]; setnames(tga_stops, "RT", "RT_TGA")
  taa_stops<-multistops_2[treatment=="Clitocine" &stop_type=="taa", c("start", "RT")]; setnames(taa_stops, "RT", "RT_TAA")
  tag_stops<-multistops_2[treatment=="Clitocine" &stop_type=="tag", c("start", "RT")]; setnames(tag_stops, "RT", "RT_TAG")
  tga_taa_stops<-full_join(full_join(tga_stops,taa_stops, by="start"), tag_stops, by="start")
  p5<-ggplot(tga_taa_stops[!is.na(RT_TAG)], aes(x=RT_TAA, y=RT_TAG)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("Clitocine") + p
  p6<-ggplot(tga_taa_stops[!is.na(RT_TGA)], aes(x=RT_TAA, y=RT_TGA)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("Clitocine")+ p
  
  tga_stops<-multistops_2[treatment=="SJ6986" &stop_type=="tga", c("start", "RT")]; setnames(tga_stops, "RT", "RT_TGA")
  taa_stops<-multistops_2[treatment=="SJ6986" &stop_type=="taa", c("start", "RT")]; setnames(taa_stops, "RT", "RT_TAA")
  tag_stops<-multistops_2[treatment=="SJ6986" &stop_type=="tag", c("start", "RT")]; setnames(tag_stops, "RT", "RT_TAG")
  tga_taa_stops<-full_join(full_join(tga_stops,taa_stops, by="start"), tag_stops, by="start")
  p7<-ggplot(tga_taa_stops, aes(x=RT_TAA, y=RT_TAG)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("SJ6986") + p
  p8<-ggplot(tga_taa_stops[!is.na(RT_TGA)], aes(x=RT_TAA, y=RT_TGA)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("SJ6986")  + p
  
  tga_stops<-multistops_2[treatment=="G418" &stop_type=="tga", c("start", "RT")]; setnames(tga_stops, "RT", "RT_TGA")
  taa_stops<-multistops_2[treatment=="G418" &stop_type=="taa", c("start", "RT")]; setnames(taa_stops, "RT", "RT_TAA")
  tag_stops<-multistops_2[treatment=="G418" &stop_type=="tag", c("start", "RT")]; setnames(tag_stops, "RT", "RT_TAG")
  tga_taa_stops<-full_join(full_join(tga_stops,taa_stops, by="start"), tag_stops, by="start")
  p9<-ggplot(tga_taa_stops[!is.na(RT_TAG)], aes(x=RT_TAA, y=RT_TAG)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("G418") + p
  p10<-ggplot(tga_taa_stops[!is.na(RT_TGA)], aes(x=RT_TAA, y=RT_TGA)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("G418")  + p
  
  
  tga_stops<-multistops_2[treatment=="CC90009" &stop_type=="tga", c("start", "RT")]; setnames(tga_stops, "RT", "RT_TGA")
  taa_stops<-multistops_2[treatment=="CC90009" &stop_type=="taa", c("start", "RT")]; setnames(taa_stops, "RT", "RT_TAA")
  tag_stops<-multistops_2[treatment=="CC90009" &stop_type=="tag", c("start", "RT")]; setnames(tag_stops, "RT", "RT_TAG")
  tga_taa_stops<-full_join(full_join(tga_stops,taa_stops, by="start"), tag_stops, by="start")
  p11<-ggplot(tga_taa_stops[!is.na(RT_TAG)], aes(x=RT_TAA, y=RT_TAG)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("CC90009") + p
  p12<-ggplot(tga_taa_stops[!is.na(RT_TGA)], aes(x=RT_TAA, y=RT_TGA)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("CC90009")  + p + scale_x_continuous(breaks=c(.5,1))
  
  tga_stops<-multistops_2[treatment=="FUr" &stop_type=="tga", c("start", "RT")]; setnames(tga_stops, "RT", "RT_TGA")
  taa_stops<-multistops_2[treatment=="FUr" &stop_type=="taa", c("start", "RT")]; setnames(taa_stops, "RT", "RT_TAA")
  tag_stops<-multistops_2[treatment=="FUr" &stop_type=="tag", c("start", "RT")]; setnames(tag_stops, "RT", "RT_TAG")
  tga_taa_stops<-full_join(full_join(tga_stops,taa_stops, by="start"), tag_stops, by="start")
  p13<-ggplot(tga_taa_stops[!is.na(RT_TAG)], aes(x=RT_TAA, y=RT_TAG)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("FUr") + p
  p14<-ggplot(tga_taa_stops[!is.na(RT_TGA)], aes(x=RT_TAA, y=RT_TGA)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("FUr")  + p
  
  tga_stops<-multistops_2[treatment=="Gentamicin" &stop_type=="tga", c("start", "RT")]; setnames(tga_stops, "RT", "RT_TGA")
  taa_stops<-multistops_2[treatment=="Gentamicin" &stop_type=="taa", c("start", "RT")]; setnames(taa_stops, "RT", "RT_TAA")
  tag_stops<-multistops_2[treatment=="Gentamicin" &stop_type=="tag", c("start", "RT")]; setnames(tag_stops, "RT", "RT_TAG")
  tga_taa_stops<-full_join(full_join(tga_stops,taa_stops, by="start"), tag_stops, by="start")
  p15<-ggplot(tga_taa_stops[!is.na(RT_TAG)], aes(x=RT_TAA, y=RT_TAG)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("Gentamicin") + scale_x_continuous(breaks=seq(0.16, 0.28, 0.06)) + p
  p16<-ggplot(tga_taa_stops[!is.na(RT_TGA)], aes(x=RT_TAA, y=RT_TGA)) + geom_jitter(position=position_jitter(0.01)) + geom_smooth(method='lm', se = FALSE) + stat_cor(aes(label = ..r.label..), size=4) + ggtitle("Gentamicin") + scale_x_continuous(breaks=seq(0.15, 0.4, 0.1)) + p
  
  
  p17a<-grid.arrange(p1, p3, p5, p7, p9, p11, p13, p15, ncol=8, left = textGrob("TAG", rot = 90, vjust = .5, gp = gpar(fontsize = 17)))
  p17b<-grid.arrange(p2, p4, p6, p8, p10, p12, p14, p16, ncol=8, left = textGrob("TGA", rot = 90, vjust = .5, gp = gpar(fontsize = 17)))
  
  p17<-grid.arrange(p17a, p17b, ncol=1, bottom = textGrob("TAA", gp = gpar(fontsize = 17)))
  p17
```
