---
title: "Dimsum"
output: html_document
---

**Dimsum generates a read counts table for all variants across drugs from FastQ files**.

**Drugs were tested in two different rounds (named 'A' and 'B') and we did a separate Dimsum run for each**.

**Both round_A and round_B read counts tables (Dimsum output) are merged together into 'treated_samples' dataset in the Generate_treated_samples.Rmd file.**

**Download FastQ files from Sequence Read Archive (SRA) with accession number PRJNA996618: http://www.ncbi.nlm.nih.gov/bioproject/996618 to your base directory (base_dir)**

- Select all FastQ files for Clitocine, DAP and SRI drugs and store them in a subdirectory from your base_dir named 'round_A'.

- Select all FastQ files for CC90009, FUr, Gentamicin, G418, SJ6986 and Untreated drugs and store them in a subdirectory from your base_dir named 'round_B'.


**For round_A dimsum run use:**

- fastqFileDir="/base_dir/round_A"

- experimentDesignPath="/base_dir/Dimsum/experimental_design_dimsum_diseasePTC_round_A.txt"

- outputPath="/base_dir/round_A_ouput"

**For round_B dimsum run use:**

- fastqFileDir="/base_dir/round_B"

- experimentDesignPath="/base_dir/Dimsum/experimental_design_dimsum_diseasePTC_round_B.txt"

- outputPath="/base_dir/round_B_ouput"

```
fastqFileDir="/base_dir/round_A"
experimentDesignPath="/base_dir/experimental_design_dimsum_diseasePTC_round_A.txt"
gzipped=F
cutadapt5First="ACCGGTACCGTCGACACG;required...TTCCTGACTGTTACGGCG;optional"
cutadapt5Second="CGCCGTAACAGTCAGGAA;required...CGTGTCGACGGTACCGGT;optional"
cutadaptErrorRate="0.2"
cutadaptMinLength="90"
cutadaptOverlap="3"
outputPath="/base_dir/round_A_ouput"
projectName="dimsum_diseasePTC"
wildtypeSequence="CTGTCTCACCGCAGGGCCTCCTTGGAGGAGAGAGCCGTTTCCAGAAACCCCCTGTGCACCCTGTGCCTGGAGTAGCGCAGGCACCCAACAGCCACGCCCTGCGGCCACCTGTTCTGCTGGGAGTGCATCACCGCGTGGTGCAGCAGC"
barcodeIdentityPath="/base_dir/VariantIdentity_diseasePTC.txt"
sequenceType="noncoding"
startStage=0
stopStage=5
vsearchMinovlen=10
numCores=4
retainItermediateFiles=T
vsearchMinQual="10"
mutagenesisType="random"
maxSubstitutions=200
indels="all"

DiMSum -i $fastqFileDir -e $experimentDesignPath --vsearchMinQual $vsearchMinQual --retainIntermediateFiles $retainItermediateFiles --vsearchMinovlen $vsearchMinovlen --num
Cores $numCores --cutadapt5First $cutadapt5First --gzipped $gzipped -\
-cutadapt5Second $cutadapt5Second -a $cutadaptErrorRate --cutadaptMinLength $cutadaptMinLength --cutadaptOverlap $cutadaptOverlap -o $outputPath -p $projectName -w $wildtyp
eSequence --startStage $startStage --stopStage $stopStage --mutagenes\
isType $mutagenesisType --barcodeIdentityPath $barcodeIdentityPath --sequenceType $sequenceType --maxSubstitutions $maxSubstitutions --indels $indels
```
